stdlib/public/TensorFlow/Dataset.swift

```diff
@@ -0,0 +1,141 @@
+//===-- Dataset.swift -----------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// The dataset API.
+//
+//===----------------------------------------------------------------------===//
+
+/// Represents a potentially large set of elements.
+///
+/// A `SingleValueDataset` can be used to represent an input pipeline as a
+/// collection of element tensors.
+@_fixed_layout
+public struct SingleValueDataset {
+  // NOTE: SingleValueDataset is currently specialzed on `Float` to work around
+  // deabstraction bug SR-8463. This is to avoid indirect passing in SIL. When
+  // the bug gets resolved, `SingleValueDataset` will have generic parameter
+  // `Scalar`.
+  public typealias Scalar = Float
+  @usableFromInline let _handle: VariantHandle
+  public let elementShape: TensorShape
+
+  @usableFromInline @inline(__always)
+  internal init(_handle: VariantHandle, elementShape: TensorShape) {
+    self._handle = _handle
+    self.elementShape = elementShape
+  }
+}
+
+public extension SingleValueDataset {
+  /// Creates a dataset from a batch of elements as a tensor.
+  ///
+  /// - Parameter
+  ///   - elements: The batch of elements.
+  ///   - elementShape: The shape that the 
+  ///
+  // FIXME: Due to inability to perform shape inference during deabstraction,
+  // users must specify the element shape for today.
+  @inlinable @inline(__always)
+  public init(elements: Tensor<Scalar>, elementShape: TensorShape) {
+    self.init(
+      _handle: #tfop(
+        "TensorSliceDataset", [elements],
+        Toutput_types: [Scalar.self],
+        output_shapes: [elementShape]
+      ),
+      elementShape: elementShape
+    )
+  }
+}
+
+extension SingleValueDataset : Sequence {
+  public typealias Element = Tensor<Scalar>
+
+  public typealias Iterator = SingleValueDatasetIterator
+
+  /// Returns an iterator over the elements of this dataset.
+  @inlinable @inline(__always)
+  public func makeIterator() -> SingleValueDatasetIterator {
```

Comment:
another design option is to avoid exposing the "cardinality" dimension of the dataset (e.g. SingleValueDataset, DoubleValueDataset etc) into different Dataset structs, and instead extend makeIterator() with param(s) to indicate the cardinality of the iterator we create.

This way when we later of datasets that read from different sources (files, etc), we won't need a corss-product of dataset structs from {single, double, tripple value ... } X {data set from tensor, data set from file, etc).

 WDYT?

Reply:
@mhong What do you mean exactly by "extending makeIterator() with param(s)"?
The `Sequence` protocols requires a [`makeIterator()`](https://developer.apple.com/documentation/swift/sequence/2885155-makeiterator) method that takes no arguments.

Eventually, I think the true way to generalize over `n` outputs is variadic generics.

Reply:
Why do we need to conform dataset to `Sequence` protocol?

Reply:
`for-in` loops work on types that conform to `Sequence`.
This allows us to write things like `for item in dataset { ... }`, as you can see in the tests.

Reply:
I have explanation in the PR description. Please take a read. It is a type system limitation.

Reply:
> extend makeIterator() with param(s) to indicate the cardinality of the iterator we create.

This is not a viable design. A Dataset and an Iterator have the same output nesting structure, like “cardinality” you referred to.

Reply:
> Eventually, I think the true way to generalize over `n` outputs is variadic generics. 

Dan is correct. Variadic generics are the only way to support it natively. It’s not possible in Swift today.

Reply:
I don't understand the exchange here.  I think that Mingsheng has the right intuition - we shouldn't have one Swift type for each dataset element count that becomes interesting in the future.  We don't have the mechanics to support this at the moment, but we should build towards it.  I don't think we need variadic tuples, we just need to eventually have something like:

struct DataSet<T> : Sequence {
  typealias Element = T
...
}

Then you could use DataSet<Tensor<Float>> for a single value dataset.  If you want to support multiple value datasets, you'd define a struct like:

struct MyDataSetRecord {
  var x : Tensor<Float>
  var y : Tensor<Int8>
}

and use DataSet<MyDataSetRecord>.  I think this should compose together naturally, but we are missing the support for #tfop type inference to handle returns of structs like MyDataSetRecord.  If we handled them as multiple returns of the corresponding tensor elements, then it should all "just work" with the existing dataset #tfop's AFAIK.

Reply:
> I don't understand the exchange here. I think that Mingsheng has the right intuition - we shouldn't have one Swift type for each dataset element count that becomes interesting in the future.

Using the single type instead of multiple types was my first consideration. I agree with your intuition as well, and I also dislike using specialized types for datasets with different result values. However, it won't work in today's `#tfop` because one can't turn an aggregate into a list of metatypes of tensor dtypes to pass in dataset ops.

Moreover, it is questionable whether `#tfop` can be extended to support dataset op requirements. Element shapes and element dtypes are required by dataset ops. How would you store element shapes such that they can be passed as an array? And how would you do so for scalar metatypes? 

In your `MyDataSetRecord` case, we'd need to
1) constrain all elements to `Tensor` (this could be done via protocol conformance and derived conformances, but it's very complex for this purpose),

2) extract the metatype from each field's associated `Scalar` type and pass them as a const array to dataset tfops (this is impossible today and it would complicate `#tfop` a lot), and

3) either have a struct parallel to `MyDataSetRecord` to store all element shapes, or store an element shape as part of the `MyDataSetRecord`.
    ```swift
    struct MyDataSetRecord {
      var x: (Tensor<Float>, TensorShape)
      var y: (Tensor<Int8>, TensorShape)
    }
    ```
    This would again make `#tfop` very complicated for extracting struct members. 

> If we handled them as multiple returns of the corresponding tensor elements, then it should all "just work" with the existing dataset #tfop's AFAIK.

As shown above, it is not just extending `#tfop` to support multiple returns, but rather extending `#tfop` to support a form of operations analogous to nest mapping. And that seems very far away from today's `#tfop` design, and the complexity is unmeasured.